### ISSUE TITLE
Type checking for **kwds based on its use

### ISSUE THREAD

**Issue by mauvilsa** [microsoft/pyright#3583]
**Is your feature request related to a problem? Please describe.**
Type checking should work when `**kwds` does not have a type annotation but what it accepts can be deduced based on its use.

**Describe the solution you'd like**
Look at the following example:

```python
def some_function(f1: int):
    assert isinstance(f1, int)

def other_function(f0: str, **kwds):
    assert isinstance(f0, str)
    some_function(**kwds)

other_function(f0='a', f1='b')
```

I would expect a static type checker to warn that `f1='b'` is wrong because the type should be `int`. There shouldn't be a need to add a type to `**kwds` since what is accepted can be deduced from how it is used. Note that `some_function` could be defined in another module and be used in multiple places, thus the "don't repeat yourself" and "separation of concerns" principles apply. Better to have the type for `f1` be only in the definition of `some_function` and not in the `**kwds`.

**Additional context**
I created a similar issue in mypy https://github.com/python/mypy/issues/12991

I have been developing [jsonargparse](github.com/omni-us/jsonargparse) which can be used for automatically creating complex parsers based on [signatures](https://jsonargparse.readthedocs.io/en/stable/#classes-methods-and-functions). Figuring out the accepted keyword args and its types based on how `**kwds` is used, to me seems like a natural feature to expect. jsonargparse 4.10.0.dev2 already supports many cases like this, see [AST resolver](https://jsonargparse.readthedocs.io/en/latest/#ast-resolver). I just wonder why static type checkers are not going in this direction.

**Comment by erictraut** [comment]
That's not how static type checking works for functions. PEP 484 is very clear that if a parameter has no type annotation, it is meant to be interpreted as an implicit `Any` type. 

Pyright has a "strict" type checking mode that will report an error if a parameter is unannotated, since this is a potential source of errors. If you provide an appropriate annotation for `**kwargs`, pyright will then validate the type according to the type rules set forth in PEP 484.

**Comment by mauvilsa** [comment]
@erictraut thank you very much for the fast response. I understand that pyright strictly follows PEP 484 and my intention is not to suggest that it shouldn't. Though I don't see PEP 484 explicitly saying that what `**kwds` accepts, cannot be deduced from its use. From my understanding to point of a static type checker it to identify errors in the code. In the snippet above it can be clearly seen that there is a mistake, the kind of which is possible to warn about. The fact that it doesn't just looks like a lack of a feature from a perspective of a pyright's user. And as I mentioned, adding a type `**kwds` goes against software engineering principles.

Is your belief that I should push for there being a PEP that explicitly addresses this topic? Would that be just creating an issue in https://github.com/python/peps/issues?

**Comment by erictraut** [comment]
You are certainly welcome to start a discussion about this in the typing-sig or submit a PEP proposal, but I think it will be very unlikely to gain support. What you're suggesting here is inconsistent with existing static typing rules in Python, and it would be a backward compatibility problem for existing typed code bases that assume existing PEP 484 rules.

If you want the type of `**kwargs` to be enforced, you should include a type annotation for it, just like any other parameter.

**Comment by erictraut** [comment]
You may find [PEP 692](https://github.com/python/peps/pull/2620/files), which is in draft form and currently under review, of interest. It allows `**kwargs` parameters to be annotated with unpacked `TypedDict` types. Pyright includes provisional support for this functionality.

**Comment by mauvilsa** [comment]
Thank you for the info. I have seen the discussions about `TypedDict`, and I completely agree there is a need to allow more precise typing of `**kwargs`. However, that does not mean that in all cases `**kwargs` should be typed, (for the moment giving more importance to software engineering principles than PEPs).
